File: backend/app/schemas/order.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class AddressIn(BaseModel):
    full_name: str = Field(min_length=1, max_length=200)
    line1: str = Field(min_length=1, max_length=200)
    line2: str | None = Field(default=None, max_length=200)
    city: str = Field(min_length=1, max_length=100)
    state: str | None = Field(default=None, max_length=100)
    postal_code: str = Field(min_length=1, max_length=40)
    country: str = Field(default="US", min_length=2, max_length=2)
    phone: str | None = Field(default=None, max_length=40)


class OrderItemIn(BaseModel):
    product_id: int
    quantity: int = Field(ge=1)


class CreateOrderRequest(BaseModel):
    items: list[OrderItemIn] = Field(min_length=1)
    shipping_address: AddressIn


class OrderItemOut(BaseModel):
    id: int
    product_id: int
    title: str
    unit_price_cents: int
    quantity: int
    line_total_cents: int


class OrderOut(BaseModel):
    id: int
    buyer_id: int
    seller_id: int
    status: str
    currency: str
    subtotal_cents: int
    total_cents: int
    created_at: datetime
    items: list[OrderItemOut]

File: apps/flutter_app/lib/models/order.dart
class OrderItem {
  OrderItem({
    required this.id,
    required this.productId,
    required this.title,
    required this.unitPriceCents,
    required this.quantity,
    required this.lineTotalCents,
  });

  final int id;
  final int productId;
  final String title;
  final int unitPriceCents;
  final int quantity;
  final int lineTotalCents;

  static OrderItem fromJson(Map<String, dynamic> json) {
    return OrderItem(
      id: json['id'] as int,
      productId: json['product_id'] as int,
      title: json['title'] as String,
      unitPriceCents: json['unit_price_cents'] as int,
      quantity: json['quantity'] as int,
      lineTotalCents: json['line_total_cents'] as int,
    );
  }
}

class Order {
  Order({
    required this.id,
    required this.status,
    required this.currency,
    required this.subtotalCents,
    required this.totalCents,
    required this.items,
  });

  final int id;
  final String status;
  final String currency;
  final int subtotalCents;
  final int totalCents;
  final List<OrderItem> items;

  static Order fromJson(Map<String, dynamic> json) {
    final itemsJson = (json['items'] as List<dynamic>? ?? const []);
    return Order(
      id: json['id'] as int,
      status: json['status'] as String,
      currency: json['currency'] as String,
      subtotalCents: json['subtotal_cents'] as int,
      totalCents: json['total_cents'] as int,
      items: itemsJson
          .whereType<Map<String, dynamic>>()
          .map(OrderItem.fromJson)
          .toList(growable: false),
    );
  }
}

